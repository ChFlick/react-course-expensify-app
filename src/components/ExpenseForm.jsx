import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        const expense = props.expense || {};

        this.state = {
            description: expense.description || '',
            note: expense.note || '',
            amount: expense.amount ? (expense.amount / 100).toString() : '',
            createdAt: moment(expense.createdAt) || moment(),
            calendarFocused: false,
            error: null,
        };
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState((state) => ({ description }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d+((\.|,)\d{0,2})?$/)) {
            this.setState((state) => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState({ createdAt });
        }
    }

    onCalendarFocusChange = ({ focused }) => {
        this.setState({ calendarFocused: focused });
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState((state) => ({ note }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState({ error: 'Please provide description and amount.' });
        } else {
            this.setState({ error: null });
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount.replace(',', '.')) * 100,
                createdAt: this.state.createdAt.valueOf(),
            });
        }
    };

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onCalendarFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    displayFormat='DD.MM.YYYY'
                />
                <textarea
                    placeholder="Add a note for your expense (optional)"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >

                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        );
    }
};
