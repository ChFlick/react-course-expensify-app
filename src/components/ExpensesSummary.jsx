import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import 'numeral/locales/de';
import { Link } from 'react-router-dom';

import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

numeral.locale('de');

export const ExpensesSummary = ({ expenseCount, hiddenExpenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedTotal = numeral(expensesTotal / 100).format('0.0[,]00$');

    const hiddenExpenseWord = hiddenExpenseCount === 1 ? 'expense' : 'expenses';
    const hiddenExpensePronoun = hiddenExpenseCount === 1 ? 'is' : 'are';

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expenseCount} </span> {expenseWord} totalling <span>{formattedTotal}</span>
                </h1>
                <h3 className="page-header__subtitle">
                    {hiddenExpenseCount > 0 ?
                        `There ${hiddenExpensePronoun} ${hiddenExpenseCount} ${hiddenExpenseWord} hidden by filters.` :
                        'All expenses visible.'
                    }
                </h3>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

const matchStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        hiddenExpenseCount: state.expenses.length - visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses),
    };
};

export default connect(matchStateToProps)(ExpensesSummary);
