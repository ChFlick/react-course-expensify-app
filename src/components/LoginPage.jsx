import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLoginWithGoogle, startLoginWithGithub }) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify</h1>
            <p>It&#39;s time to get your expenses under control.</p>
            <button className="button" onClick={startLoginWithGoogle}>Login with Google</button>
            <button className="button" onClick={startLoginWithGithub}>Login with GitHub</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLoginWithGoogle: () => dispatch(startLogin('google')),
    startLoginWithGithub: () => dispatch(startLogin('github'))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
