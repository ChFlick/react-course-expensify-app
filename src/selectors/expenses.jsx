import moment from 'moment';

// Get visible expenses
export default (expenses, { text = '', sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAt = moment(expense.createdAt);

        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAt, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAt, 'day') : true;
        const textMatch = !expense.description || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        const sortField = (() => {
            if (sortBy === 'amount') return 'amount';
            if (sortBy === 'date') return 'createdAt';
        })();

        return b[sortField] - a[sortField];
    });
};
