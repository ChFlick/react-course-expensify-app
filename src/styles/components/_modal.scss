.ReactModal__Overlay--after-open {
    align-items: center;
    display: flex;
    justify-content: center;
}

.modal {
    background-color: $off-white;
    border-radius: 5px;
    max-width: 22rem;
    outline: none;
    padding: $m-size;
}

.modal__content {
    font-size: $font-size-large;
    font-weight: 400;
    margin-bottom: $m-size;
}

.modal__buttons {
    display: flex;
    justify-content: space-between;
}