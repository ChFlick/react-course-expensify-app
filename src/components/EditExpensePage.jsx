import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    state = {
        modalIsOpen: false
    };

    onEdit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    openModal = () => {
        this.setState(() => ({
            modalIsOpen: true
        }));
    };

    closeModal = () => {
        this.setState(() => ({
            modalIsOpen: false
        }));
    };

    removeExpense = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.closeModal();
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onEdit}
                    />
                    <button className="button button--secondary" onClick={this.openModal}>
                        Remove Expense
                    </button>

                    <Modal
                        isOpen={this.state.modalIsOpen}
                        onRequestClose={this.closeModal}
                        contentLabel="Remove Expense Modal"
                        className="modal"
                    >
                        <div className="modal__content">
                            Do you really want to remove the expense?
                        </div>
                        <div className="modal__buttons">
                            <button className="button" onClick={this.closeModal}>Cancel</button>
                            <button className="button" onClick={this.removeExpense}>Remove</button>
                        </div>
                    </Modal>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
