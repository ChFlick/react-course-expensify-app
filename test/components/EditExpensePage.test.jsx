import React from 'react';
import { shallow } from 'enzyme';

import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../src/components/EditExpensePage';

const expense = expenses[0];
let startEditExpense;
let startRemoveExpense;
let history;
let wrapper;

beforeEach(() => {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage
            expense={expense}
            startEditExpense={startEditExpense}
            startRemoveExpense={startRemoveExpense}
            history={history}
        />
    );
});

test('should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense);

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expense.id, expense);
});

test('should handle remove expense with modal -> remove', () => {
    wrapper.find('button').at(0).simulate('click');
    wrapper.find('button').at(2).simulate('click');

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id: expense.id});
});

test('should handle remove expense with modal -> cancel', () => {
    wrapper.find('button').at(0).simulate('click');
    wrapper.find('button').at(1).simulate('click');

    expect(history.push).not.toHaveBeenCalled();
    expect(startRemoveExpense).not.toHaveBeenCalled();
});
