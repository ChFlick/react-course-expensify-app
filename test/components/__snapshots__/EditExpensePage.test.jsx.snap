// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "coffee",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      className="modal"
      closeTimeoutMS={0}
      contentLabel="Remove Expense Modal"
      isOpen={false}
      onRequestClose={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <div>
        Do you really want to remove the expense?
      </div>
      <button
        className="button"
        onClick={[Function]}
      >
        Cancel
      </button>
      <button
        className="button"
        onClick={[Function]}
      >
        Remove
      </button>
    </Modal>
  </div>
</div>
`;
